File: byssup/bfs/triangle.py
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        prev_total_row = triangle[0]
        for row in triangle[1:]:
            cur_total_row = [99999 for _ in row]
            for i, num in enumerate(row):
                if i == 0:
                    cur_total_row[0] = prev_total_row[0] + num
                elif i == len(row) - 1:
                    cur_total_row[-1] = prev_total_row[-1] + num
                else:
                    cur_total_row[i] = min(prev_total_row[i - 1] + num, prev_total_row[i] + num)
            prev_total_row = cur_total_row
        return min(prev_total_row)

if __name__ == '__main__':
    solution = Solution()
    print solution.minimumTotal([[2], [3,4], [6,5,7],[4,1,8,3]])

File: shpark/triangle.java
class Solution {
  public int minimumTotal(List<List<Integer>> triangle) {
		int size=triangle.size();
		for (int i = 0; i <size; i++) {
			for (int j = 0; j < triangle.get(i).size(); j++) {
				int cur = triangle.get(i).get(j);
				if (i > 0) {
					if(j==0){
						triangle.get(i).set(j, triangle.get(i - 1).get(j) + cur);
					}else if(j==i){
						triangle.get(i).set(j, triangle.get(i - 1).get(j - 1) + cur);
					}else{
						triangle.get(i).set(j,Math.min(triangle.get(i - 1).get(j - 1) + cur, triangle.get(i - 1).get(j) + cur));
					}
				}
			}
		}
		int min=Integer.MAX_VALUE;
		for(int i:triangle.get(size-1))min=Math.min(min, i);
		return min;
	}
}